src/frame/decompress.rs

```diff
@@ -1 +1,403 @@
+            r: rdr,
+            src: Default::default(),
+            dst: Default::default(),
+            ext_dict_offset: 0,
+            ext_dict_len: 0,
+            dst_start: 0,
+            dst_end: 0,
+            current_frame_info: None,
+            content_hasher: XxHash32::with_seed(0),
+            content_len: 0,
+        }
+    }
+
+    pub fn frame_info(&mut self) -> Option<&FrameInfo> {
+        self.current_frame_info.as_ref()
+    }
+
+    /// Gets a reference to the underlying reader in this decoder.
+    pub fn get_ref(&self) -> &R {
+        &self.r
+    }
+
+    /// Gets a mutable reference to the underlying reader in this decoder.
+    ///
+    /// Note that mutation of the stream may result in surprising results if
+    /// this decoder is continued to be used.
+    pub fn get_mut(&mut self) -> &mut R {
+        &mut self.r
+    }
+
+    fn read_frame_info(&mut self) -> Result<usize, io::Error> {
+        let mut buffer = [0u8; MAX_FRAME_INFO_SIZE];
+        match self.r.read(&mut buffer[..MIN_FRAME_INFO_SIZE])? {
+            0 => return Ok(0),
+            MIN_FRAME_INFO_SIZE => (),
+            read => self.r.read_exact(&mut buffer[read..MIN_FRAME_INFO_SIZE])?,
+        }
+        let required = FrameInfo::read_size(&buffer[..MIN_FRAME_INFO_SIZE])?;
+        if required != MIN_FRAME_INFO_SIZE {
+            self.r
+                .read_exact(&mut buffer[MIN_FRAME_INFO_SIZE..required])?;
+        }
+        let frame_info = FrameInfo::read(&buffer[..required])?;
+        if frame_info.dict_id.is_some() {
+            // Unsupported right now so it must be None
+            return Err(Error::WrongDictionary {
+                actual: frame_info.dict_id,
+                expected: None,
+            }
+            .into());
+        }
+
+        let max_block_size = frame_info.block_size.get_size();
+        let dst_size = if frame_info.block_mode == BlockMode::Linked {
+            // In linked mode we consume the output (bumping dst_start) but leave the
+            // beginning of dst to be used as a prefix in subsequent blocks.
+            // That is at least until we have at least `max_block_size + WINDOW_SIZE`
+            // bytes in dst, then we setup an ext_dict with the last WINDOW_SIZE bytes
+            // and the output goes to the beginning of dst again.
+            // Since we always want to be able to write a full block (up to max_block_size)
+            // we need a buffer with at least `max_block_size * 2 + WINDOW_SIZE` bytes.
+            max_block_size * 2 + WINDOW_SIZE
+        } else {
+            max_block_size
+        };
+        #[cfg(feature = "safe-encode")]
+        {
+            self.src.resize(max_block_size, 0);
+            self.dst.resize(dst_size, 0);
+        }
+        #[cfg(not(feature = "safe-encode"))]
+        {
+            self.src.clear();
+            self.dst.clear();
+            self.src.reserve_exact(max_block_size);
+            self.dst.reserve_exact(dst_size);
```

Comment:
reserve_exact reserves additional capacity, not total. I think the intention here was the same as in resize to ensure the total capacity?

Although this only happens when we read concatenates frames, do you know if this is common?

Reply:
Yes, the sequence of clear->reserve_exact->set_len is meant to be a faster/uninitiated version of resize(). Since they are cleared just before the reserve_exact call it behaves as total capacity to ensure set_len will be valid.

This will happen on all frames, including single frame scenarios, as the buffers sizes ultimately depend on the fields of the FrameInfo.

My guess would be that concatenated frames are relatively rare.